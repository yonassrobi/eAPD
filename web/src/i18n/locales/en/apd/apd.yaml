title: APD Overview

overview:
  instruction:
    short: >-
      Choose the federal fiscal year(s) this APD covers.

introduction:
  instruction:
    heading: Introduction
    detail: >-
      Provide a brief introduction to the state program. Include context for
      this funding request by providing information such as current size of the
      program and how the program is positioned within the state. You might also
      include a high-level look at the history and context of the program and
      the state’s future vision for it.

hit:
  instruction:
    heading: HIT Overview
    detail: >-
      Briefly summarize planned HIT-funded activities contained within this APD.

hie:
  instruction:
    heading: HIE Overview
    detail: >-
      Summarize planned HIE-funded activities within this APD.

      If there are no HIE activities, enter “No HIE activities.”


      [ol] Description of the state’s HIE approach (statewide, sub-state health
      information organization (HIO), etc.);

      Anticipated risks and mitigation strategies;

      The short and long term value proposition to providers;

      The role of state government in governance and policy-setting. [/ol]
    helpText: >-
      CMS recommends that states connect with regional CMS HITECH contacts prior
      to developing an HIE funding request. HIE funding requests are complex.
      Early consultation with CMS facilitates a more efficient approval process.

mmis:
  instruction:
    heading: MMIS Overview
    detail: >-
      Briefly summarize planned MMIS-funded activities contained within this
      APD.

      If there are no MMIS activities, enter “No MMIS activities.”
