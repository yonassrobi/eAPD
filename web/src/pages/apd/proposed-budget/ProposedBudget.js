import React from 'react';

import { Section, Subsection } from '../../../components/Section';
import CombinedActivityCosts from './CombinedActivityCosts';
import BudgetSummary from './BudgetSummary';
import Waypoint from '../../../components/ConnectedWaypoint';
import IncentivePayments from './IncentivePayments';
import QuarterlyBudgetSummary from './QuarterlyBudgetSummary';
import AlertMissingFFY from '../../../components/AlertMissingFFY';

const ProposedBudget = () => (
  <React.Fragment>
    <Waypoint />
    <AlertMissingFFY />
    <Section resource="proposedBudget">
      <Waypoint id="combined-activity-costs-table" />
      <Subsection
        id="combined-activity-costs-table"
        resource="proposedBudget.combinedActivityCosts"
      >
        <CombinedActivityCosts />
      </Subsection>

      <Waypoint id="budget-summary-table" />
      <Subsection
        id="budget-summary-table"
        resource="proposedBudget.summaryBudget"
      >
        <BudgetSummary />
      </Subsection>

      <Waypoint id="budget-federal-by-quarter" />
      <Subsection
        id="budget-federal-by-quarter"
        resource="proposedBudget.quarterlyBudget"
      >
        <QuarterlyBudgetSummary />
      </Subsection>

      <Waypoint id="budget-incentive-by-quarter" />
      <Subsection
        id="budget-incentive-by-quarter"
        resource="proposedBudget.paymentsByFFYQuarter"
      >
        <IncentivePayments />
      </Subsection>
    </Section>
  </React.Fragment>
);

export default ProposedBudget;
