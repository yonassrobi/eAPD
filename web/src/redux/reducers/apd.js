import u from 'updeep';
import { apply_patch as applyPatch } from 'jsonpatch';

import {
  contractorDefaultHourly,
  contractorDefaultYear,
  costAllocationEntry,
  costAllocationNarrative,
  expenseDefaultYear,
  newActivity,
  newContractor,
  newExpense,
  newMilestone,
  newOutcome,
  newOutcomeMetric,
  newStatePerson,
  quarterlyFFPEntry,
  statePersonDefaultYear
} from './activities';
import {
  ADD_APD_ITEM,
  ADD_APD_YEAR,
  EDIT_APD,
  EDIT_APD_NAME,
  REMOVE_APD_ITEM,
  REMOVE_APD_YEAR
} from '../actions/editApd';
import {
  CREATE_APD_SUCCESS,
  FETCH_ALL_APDS_FAILURE,
  FETCH_ALL_APDS_REQUEST,
  FETCH_ALL_APDS_SUCCESS,
  RESET,
  SAVE_APD_SUCCESS,
  SELECT_APD_SUCCESS,
  SELECT_APD_FAILURE,
  SET_APD_TO_SELECT_ON_LOAD,
  ADMIN_CHECK_TOGGLE,
  ADMIN_CHECK_COLLAPSE_TOGGLE,
  ADMIN_CHECK_COMPLETE_TOGGLE
} from '../actions/app';
import { generateKey, defaultAPDYearOptions } from '@cms-eapd/common';
import initialAssurances from '../../util/regulations';

export const getPatchesToAddYear = (state, year) => {
  const years = [...state.data.years, year].sort();
  const patches = [
    { op: 'replace', path: '/years', value: years },
    {
      op: 'add',
      path: `/proposedBudget/incentivePayments/ehAmt/${year}`,
      value: { 1: 0, 2: 0, 3: 0, 4: 0 }
    },
    {
      op: 'add',
      path: `/proposedBudget/incentivePayments/ehCt/${year}`,
      value: { 1: 0, 2: 0, 3: 0, 4: 0 }
    },
    {
      op: 'add',
      path: `/proposedBudget/incentivePayments/epAmt/${year}`,
      value: { 1: 0, 2: 0, 3: 0, 4: 0 }
    },
    {
      op: 'add',
      path: `/proposedBudget/incentivePayments/epCt/${year}`,
      value: { 1: 0, 2: 0, 3: 0, 4: 0 }
    }
  ];

  state.data.activities.forEach((activity, activityIndex) => {
    activity.contractorResources.forEach((_, i) => {
      patches.push({
        op: 'add',
        path: `/activities/${activityIndex}/contractorResources/${i}/hourly/${year}`,
        value: contractorDefaultHourly()
      });
      patches.push({
        op: 'add',
        path: `/activities/${activityIndex}/contractorResources/${i}/years/${year}`,
        value: contractorDefaultYear()
      });
    });

    activity.expenses.forEach((_, i) => {
      patches.push({
        op: 'add',
        path: `/activities/${activityIndex}/expenses/${i}/years/${year}`,
        value: expenseDefaultYear()
      });
    });

    activity.statePersonnel.forEach((_, i) => {
      patches.push({
        op: 'add',
        path: `/activities/${activityIndex}/statePersonnel/${i}/years/${year}`,
        value: statePersonDefaultYear()
      });
    });

    patches.push({
      op: 'add',
      path: `/activities/${activityIndex}/costAllocation/${year}`,
      value: costAllocationEntry()
    });

    patches.push({
      op: 'add',
      path: `/activities/${activityIndex}/costAllocationNarrative/years/${year}`,
      value: costAllocationNarrative()
    });

    patches.push({
      op: 'add',
      path: `/activities/${activityIndex}/quarterlyFFP/${year}`,
      value: quarterlyFFPEntry()
    });
  });

  state.data.keyStatePersonnel.keyPersonnel.forEach((_, i) => {
    patches.push({
      op: 'add',
      path: `/keyStatePersonnel/keyPersonnel/${i}/costs/${year}`,
      value: 0
    });

    patches.push({
      op: 'add',
      path: `/keyStatePersonnel/keyPersonnel/${i}/fte/${year}`,
      value: 0
    });
  });

  return patches;
};

export const getPatchesToRemoveYear = (state, year) => {
  const index = state.data.years.indexOf(year);

  const patches = [
    { op: 'remove', path: `/years/${index}` },
    {
      op: 'remove',
      path: `/proposedBudget/incentivePayments/ehAmt/${year}`
    },
    {
      op: 'remove',
      path: `/proposedBudget/incentivePayments/ehCt/${year}`
    },
    {
      op: 'remove',
      path: `/proposedBudget/incentivePayments/epAmt/${year}`
    },
    {
      op: 'remove',
      path: `/proposedBudget/incentivePayments/epCt/${year}`
    }
  ];

  state.data.activities.forEach((activity, activityIndex) => {
    activity.contractorResources.forEach((_, i) => {
      patches.push({
        op: 'remove',
        path: `/activities/${activityIndex}/contractorResources/${i}/hourly/${year}`
      });
      patches.push({
        op: 'remove',
        path: `/activities/${activityIndex}/contractorResources/${i}/years/${year}`
      });
    });

    activity.expenses.forEach((_, i) => {
      patches.push({
        op: 'remove',
        path: `/activities/${activityIndex}/expenses/${i}/years/${year}`
      });
    });

    activity.statePersonnel.forEach((_, i) => {
      patches.push({
        op: 'remove',
        path: `/activities/${activityIndex}/statePersonnel/${i}/years/${year}`
      });
    });

    patches.push({
      op: 'remove',
      path: `/activities/${activityIndex}/costAllocation/${year}`
    });

    patches.push({
      op: 'remove',
      path: `/activities/${activityIndex}/costAllocationNarrative/years/${year}`
    });

    patches.push({
      op: 'remove',
      path: `/activities/${activityIndex}/quarterlyFFP/${year}`
    });
  });

  state.data.keyStatePersonnel.keyPersonnel.forEach((_, i) => {
    patches.push({
      op: 'remove',
      path: `/keyStatePersonnel/keyPersonnel/${i}/costs/${year}`
    });

    patches.push({
      op: 'remove',
      path: `/keyStatePersonnel/keyPersonnel/${i}/fte/${year}`
    });
  });

  return patches;
};

const getHumanDatestamp = iso8601 => {
  const d = new Date(iso8601);

  return `${d.toLocaleDateString('en-us', {
    month: 'long',
    day: 'numeric',
    year: 'numeric'
  })}`;
};

const getHumanTimestamp = iso8601 => {
  const d = new Date(iso8601);

  return `${getHumanDatestamp(iso8601)}, ${d.toLocaleTimeString('en-us', {
    timeZoneName: 'short',
    hour12: true,
    hour: 'numeric',
    minute: 'numeric'
  })}`;
};

const getKeyPersonnelSplit = (years, apdType) => {
  let split;
  switch (apdType) {
    case 'HITECH':
      split = years.reduce(
        (s, year) => ({ ...s, [year]: { federal: 90, state: 10 } }),
        {}
      );
      break;
    case 'MMIS':
      split = years.reduce(
        (s, year) => ({ ...s, [year]: { federal: 0, state: 0 } }),
        {}
      );
      break;
    default:
      return years.reduce(
        (s, year) => ({ ...s, [year]: { federal: 0, state: 0 } }),
        {}
      );
  }
  return split;
};

export const getKeyPersonnel = (
  years = [],
  isPrimary = false,
  apdType = ''
) => ({
  costs: years.reduce((c, year) => ({ ...c, [year]: 0 }), {}),
  email: '',
  expanded: true,
  hasCosts: null,
  isPrimary,
  fte: years.reduce((p, year) => ({ ...p, [year]: 0 }), {}),
  split: getKeyPersonnelSplit(years, apdType),
  medicaidShare: years.reduce((p, year) => ({ ...p, [year]: 0 }), {}),
  name: '',
  position: '',
  key: generateKey()
});

export const getAPDId = ({
  apd: {
    data: { id }
  }
}) => id;

export const getAPDCreation = ({
  apd: {
    data: { created }
  }
}) => created;

export const getAPDName = ({
  apd: {
    data: { name }
  }
}) => name;

export const getAPDYearRange = ({
  apd: {
    data: { years }
  }
}) =>
  years && years.length
    ? `${years[0]}${years.length > 1 ? `-${years[years.length - 1]}` : ''}`
    : '';

export const getPatchesForAddingItem = (state, path, key = null) => {
  switch (path) {
    case '/keyStatePersonnel/keyPersonnel/-':
      return [
        {
          op: 'add',
          path,
          value: getKeyPersonnel(
            state.data.years,
            !state.data.keyStatePersonnel?.keyPersonnel ||
              state.data.keyStatePersonnel.keyPersonnel.length === 0
          )
        }
      ];
    case '/activities/-':
      return [
        {
          op: 'add',
          path,
          value: newActivity({
            years: state.data.years,
            key,
            apdType: state.data.apdType
          })
        }
      ];
    default:
      if (/^\/activities\/\d+\/contractorResources\/-$/.test(path)) {
        return [{ op: 'add', path, value: newContractor(state.data.years) }];
      }
      if (/^\/activities\/\d+\/expenses\/-$/.test(path)) {
        return [{ op: 'add', path, value: newExpense(state.data.years) }];
      }
      if (/^\/activities\/\d+\/outcomes\/-$/.test(path)) {
        return [{ op: 'add', path, value: newOutcome() }];
      }
      if (/^\/activities\/\d+\/outcomes\/\d+\/metrics\/-$/.test(path)) {
        return [{ op: 'add', path, value: newOutcomeMetric() }];
      }
      if (/^\/activities\/\d+\/milestones\/-$/.test(path)) {
        return [{ op: 'add', path, value: newMilestone() }];
      }
      if (/^\/activities\/\d+\/statePersonnel\/-$/.test(path)) {
        return [{ op: 'add', path, value: newStatePerson(state.data.years) }];
      }
      return [{ op: 'add', path, value: null }];
  }
};

const getAssurancesAndCompliances = assurancesAndCompliances =>
  Object.values(assurancesAndCompliances).every(regs => regs.length > 0)
    ? assurancesAndCompliances
    : initialAssurances;

const initialState = {
  data: {},
  byId: {},
  fetching: false,
  loaded: false,
  error: '',
  selectAPDOnLoad: false,
  adminCheck: {
    errors: [],
    enabled: false,
    collapsed: false,
    complete: false
  }
};

// eslint-disable-next-line default-param-last
const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_APD_ITEM: {
      const patches = getPatchesForAddingItem(state, action.path, action.key);
      return {
        ...state,
        data: applyPatch(state.data, patches)
      };
    }

    case ADD_APD_YEAR: {
      const patches = getPatchesToAddYear(state, action.value);
      return { ...state, data: applyPatch(state.data, patches) };
    }

    case EDIT_APD: {
      return {
        ...state,
        data: applyPatch(state.data, [
          {
            op: 'replace',
            path: action.path,
            value: action.value
          }
        ])
      };
    }

    case EDIT_APD_NAME: {
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.id]: {
            ...state.byId[action.id],
            name: action.name
          }
        }
      };
    }

    case REMOVE_APD_ITEM: {
      return {
        ...state,
        data: applyPatch(state.data, [
          {
            op: 'remove',
            path: action.path
          }
        ])
      };
    }

    case REMOVE_APD_YEAR: {
      const patches = getPatchesToRemoveYear(state, action.value);
      return { ...state, data: applyPatch(state.data, patches) };
    }

    case ADMIN_CHECK_TOGGLE: {
      return {
        ...state,
        adminCheck: {
          ...state.adminCheck,
          enabled: action.data
        }
      };
    }

    case ADMIN_CHECK_COLLAPSE_TOGGLE: {
      return {
        ...state,
        adminCheck: {
          ...state.adminCheck,
          collapsed: action.data
        }
      };
    }

    case ADMIN_CHECK_COMPLETE_TOGGLE: {
      return {
        ...state,
        adminCheck: {
          ...state.adminCheck,
          complete: action.data
        }
      };
    }

    case CREATE_APD_SUCCESS:
      return u(
        {
          byId: {
            [action.data.id]: {
              ...action.data,
              created: getHumanDatestamp(action.data.created),
              updated: getHumanTimestamp(action.data.updated),
              yearOptions: defaultAPDYearOptions()
            }
          }
        },
        state
      );
    case FETCH_ALL_APDS_REQUEST:
      return { ...state, fetching: true, error: '' };
    case FETCH_ALL_APDS_SUCCESS: {
      return {
        ...state,
        fetching: false,
        loaded: true,
        byId: action.data.reduce(
          (acc, apd) => ({
            ...acc,
            [apd.id]: {
              ...apd,
              created: getHumanDatestamp(apd.created),
              updated: getHumanTimestamp(apd.updated)
            }
          }),
          {}
        )
      };
    }

    case FETCH_ALL_APDS_FAILURE:
      return { ...state, fetching: false, error: action.error };

    case SAVE_APD_SUCCESS:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.data.apd.id]: {
            ...state.byId[action.data.apd.id],
            updated: getHumanTimestamp(action.data.apd.updated)
          }
        },
        data: {
          ...state.data,
          created: getHumanDatestamp(action.data.apd.created),
          updated: getHumanTimestamp(action.data.apd.updated)
        },
        adminCheck: {
          ...state.adminCheck,
          errors: action.data.adminCheck
        }
      };

    case RESET:
      return { ...state, data: {} };

    case SELECT_APD_SUCCESS: {
      return {
        ...state,
        data: {
          ...action.data.apd,
          created: getHumanDatestamp(action.data.apd.created),
          updated: getHumanTimestamp(action.data.apd.updated),
          keyStatePersonnel: {
            ...action.data.apd.keyStatePersonnel,
            keyPersonnel: action.data.apd.keyStatePersonnel.keyPersonnel.map(
              kp => ({
                ...kp,
                key: generateKey()
              })
            )
          },
          activities: action.data.apd.activities.map(
            ({
              activityId,
              milestones,
              outcomes,
              statePersonnel,
              expenses,
              contractorResources,
              ...activity
            }) => ({
              ...activity,
              key: activityId,
              activityId,
              milestones: milestones.map(milestone => ({
                ...milestone,
                key: generateKey()
              })),
              outcomes: outcomes.map(({ metrics = [], ...outcome }) => ({
                ...outcome,
                metrics: metrics.map(metric => ({
                  ...metric,
                  key: generateKey()
                })),
                key: generateKey()
              })),
              statePersonnel: statePersonnel.map(person => ({
                ...person,
                key: generateKey()
              })),
              expenses: expenses.map(expense => ({
                ...expense,
                key: generateKey()
              })),
              contractorResources: contractorResources.map(contractor => ({
                ...contractor,
                key: generateKey()
              }))
            })
          ),
          assurancesAndCompliances: getAssurancesAndCompliances(
            action.data.apd.assurancesAndCompliances
          )
        },
        adminCheck: {
          ...state.adminCheck,
          errors: action.data.adminCheck
        }
      };
    }
    case SELECT_APD_FAILURE:
      return { ...state, data: {}, fetching: false, error: action.data };
    case SET_APD_TO_SELECT_ON_LOAD:
      return { ...state, selectAPDOnLoad: true };

    default:
      return state;
  }
};

export default reducer;
