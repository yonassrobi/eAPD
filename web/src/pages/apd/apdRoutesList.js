import React from 'react';

import Activities from './activities/activities-dashboard/ActivitiesDashboard';
import AssurancesAndCompliance from './assurances-and-compliance/AssurancesAndCompliance';
import StatePrioritiesAndScope from './state-priorities-and-scope/StatePrioritiesAndScope';
import Export from './export/ApdExport';
import ApdOverview from './apd-overview/ApdOverview';
import ExecutiveSummary from './executive-summary/ExecutiveSummary';
import PreviousActivities from './previous-activities/PreviousActivities';
import ProposedBudget from './proposed-budget/ProposedBudget';
import ScheduleSummary from './schedule-summary/ScheduleSummary';
import SecurityPlanning from './security-planning/SecurityPlanning';
import KeyStatePersonnel from './key-state-personnel/KeyStatePersonnel';

const apdOverview = {
  path: '/apd/:apdId/apd-overview',
  children: <ApdOverview />,
  isPublic: false,
  contentType: 'form',
  siteSection: 'APD',
  pageName: 'Overview'
};

const statePrioritiesAndScope = {
  path: '/apd/:apdId/state-priorities-and-scope',
  children: <StatePrioritiesAndScope />,
  isPublic: false,
  contentType: 'form',
  siteSection: 'APD',
  pageName: 'State Priorities and Scope of APD'
};

const keyStatePersonnel = {
  path: '/apd/:apdId/state-profile',
  children: <KeyStatePersonnel />,
  isPublic: false,
  contentType: 'form',
  siteSection: 'APD',
  pageName: 'Key State Personnel'
};

const previousActivities = {
  path: '/apd/:apdId/previous-activities',
  children: <PreviousActivities />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Previous Activities'
};

const activitiesDashboard = {
  path: '/apd/:apdId/activities',
  children: <Activities />,
  isPublic: false,
  contentType: 'form',
  siteSection: 'APD',
  pageName: 'Activities Dashboard'
};

const schedule = {
  path: '/apd/:apdId/schedule-summary',
  children: <ScheduleSummary />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Schedule Summary'
};

const proposedBudget = {
  path: '/apd/:apdId/proposed-budget',
  children: <ProposedBudget />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Proposed Budget'
};

const securityPlanning = {
  path: '/apd/:apdId/security-planning',
  children: <SecurityPlanning />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Security Planning'
};

const assurancesAndCompliance = {
  path: '/apd/:apdId/assurances-and-compliance',
  children: <AssurancesAndCompliance />,
  isPublic: false,
  contentType: 'form',
  siteSection: 'APD',
  pageName: 'Assurances and Compliance'
};

const executiveSummary = {
  path: '/apd/:apdId/executive-summary',
  children: <ExecutiveSummary />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Executive Summary'
};

const exportApd = {
  path: '/apd/:apdId/export',
  children: <Export />,
  isPublic: false,
  contentType: 'report',
  siteSection: 'APD',
  pageName: 'Export'
};

const routes = [
  apdOverview,
  statePrioritiesAndScope,
  keyStatePersonnel,
  previousActivities,
  securityPlanning,
  activitiesDashboard,
  schedule,
  proposedBudget,
  assurancesAndCompliance,
  executiveSummary,
  exportApd
];

export default routes;
