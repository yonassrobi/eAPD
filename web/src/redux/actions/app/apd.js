import { push } from 'connected-react-router';
import { hasBudgetUpdate } from '@cms-eapd/common';

import {
  CREATE_APD_FAILURE,
  CREATE_APD_REQUEST,
  CREATE_APD_SUCCESS,
  DELETE_APD_FAILURE,
  DELETE_APD_REQUEST,
  DELETE_APD_SUCCESS,
  FETCH_ALL_APDS_FAILURE,
  FETCH_ALL_APDS_REQUEST,
  FETCH_ALL_APDS_SUCCESS,
  SAVE_APD_FAILURE,
  SAVE_APD_REQUEST,
  SAVE_APD_SUCCESS,
  SELECT_APD_SUCCESS,
  SELECT_APD_REQUEST,
  SELECT_APD_FAILURE,
  SET_APD_TO_SELECT_ON_LOAD,
  ADMIN_CHECK_TOGGLE,
  ADMIN_CHECK_COLLAPSE_TOGGLE,
  ADMIN_CHECK_COMPLETE_TOGGLE
} from './symbols';
import { loadBudget } from '../budget';
import { APD_ACTIVITIES_CHANGE, EDIT_APD } from '../editApd/symbols';
import { alertApdCreateSuccess } from '../alert';
import {
  ariaAnnounceApdCreateSuccess,
  ariaAnnounceApdLoaded,
  ariaAnnounceApdLoading,
  ariaAnnounceApdLoadingFailure
} from '../aria';
import { t } from '../../../i18n';

import { selectApdData } from '../../selectors/apd.selectors';
import {
  selectHasChanges,
  selectPatches
} from '../../selectors/patch.selectors';
import {
  getIsFedAdmin,
  getCanUserEditAPD
} from '../../selectors/user.selector';

import axios from '../../../util/api';
import initialAssurances from '../../../util/regulations';

const LAST_APD_ID_STORAGE_KEY = 'last-apd-id';

// eslint-disable-next-line react/display-name
export const saveApd = () => (dispatch, getState) => {
  const state = getState();
  const canEdit = getCanUserEditAPD(state);
  if (canEdit) {
    const hasChanges = selectHasChanges(state);

    if (!hasChanges) {
      return Promise.resolve();
    }

    dispatch({ type: SAVE_APD_REQUEST });

    const { id: apdID } = selectApdData(state);
    const patches = selectPatches(state);

    return axios
      .patch(`/apds/${apdID}`, patches)
      .then(res => {
        const budget = (res.data.budget ??= {});

        dispatch({
          type: SAVE_APD_SUCCESS,
          data: { apd: res.data.apd, adminCheck: res.data.adminCheck }
        });
        if (hasBudgetUpdate(patches)) {
          dispatch(loadBudget(budget));
        }
        return res.data.apd;
      })
      .catch(error => {
        if (error.response && error.response.status === 403) {
          dispatch({
            type: SAVE_APD_FAILURE,
            data: t('errors.save-apd.not-logged-in')
          });
        } else {
          dispatch({ type: SAVE_APD_FAILURE });
        }
        throw error;
      });
  }
  dispatch({
    type: SAVE_APD_FAILURE,
    data: t('errors.save-apd.no-save-permissions')
  });
  return null;
};

export const selectApd =
  (id, route, { global = window, pushRoute = push } = {}) =>
  dispatch => {
    dispatch({ type: SELECT_APD_REQUEST });
    dispatch(ariaAnnounceApdLoading());

    return axios
      .get(`/apds/${id}`)
      .then(res => {
        const budget = (res.data.budget ??= {});

        dispatch({
          type: SELECT_APD_SUCCESS,
          data: { apd: res.data.apd, adminCheck: res.data.adminCheck }
        });
        dispatch({
          type: APD_ACTIVITIES_CHANGE,
          activities: res.data.apd.activities
        });
        dispatch({ type: ADMIN_CHECK_TOGGLE, data: false });

        // By default, APDs get an empty object for federal citations. The canonical list of citations is in frontend
        // code, not backend. So if we get an APD with no federal citations, set its federal citations to the initial
        // values using an EDIT_APD action. That way the initial values get saved back to the API.
        if (Object.keys(res.data.apd.assurancesAndCompliances).length === 0) {
          dispatch({
            type: EDIT_APD,
            path: '/assurancesAndCompliances',
            value: initialAssurances
          });
        }

        dispatch(loadBudget(budget));
        dispatch(pushRoute(route));
        dispatch(ariaAnnounceApdLoaded());

        if (global.localStorage) {
          global.localStorage.setItem(LAST_APD_ID_STORAGE_KEY, id);
        }
      })
      .catch(() => {
        const error =
          'This APD does not exist at this link. Check your link to make sure it is correct.';

        ariaAnnounceApdLoadingFailure(error);
        dispatch({ type: SELECT_APD_FAILURE, data: error });
        dispatch(pushRoute('/'));
      });
  };

export const setApdToSelectOnLoad = () => (dispatch, getState) => {
  const isAdmin = getIsFedAdmin(getState());
  if (!isAdmin) {
    dispatch({ type: SET_APD_TO_SELECT_ON_LOAD });
  }
};

export const createApd =
  (values, { pushRoute = push } = {}) =>
  dispatch => {
    dispatch({ type: CREATE_APD_REQUEST });
    return axios
      .post('/apds', values)
      .then(async req => {
        let data = req.data;
        dispatch({ type: CREATE_APD_SUCCESS, data: data });
        dispatch(alertApdCreateSuccess(data.id));
        dispatch(ariaAnnounceApdCreateSuccess());
        await dispatch(selectApd(data.id, `/apd/${data.id}`, { pushRoute }));
      })
      .catch(error => {
        const reason = error.response ? error.response.data : 'N/A';
        dispatch({ type: CREATE_APD_FAILURE, data: reason });
      });
  };

export const fetchAllApds =
  ({ global = window, pushRoute = push, select = selectApd } = {}) =>
  (dispatch, getState) => {
    dispatch({ type: FETCH_ALL_APDS_REQUEST });

    const url = `/apds`;

    return axios
      .get(url)
      .then(async req => {
        const apd = Array.isArray(req.data) ? req.data : null;
        dispatch({
          type: FETCH_ALL_APDS_SUCCESS,
          data: apd.filter(({ status }) => status !== 'archived')
        });

        const {
          apd: { selectAPDOnLoad }
        } = getState();
        if (
          selectAPDOnLoad &&
          global.localStorage &&
          global.localStorage.getItem(LAST_APD_ID_STORAGE_KEY)
        ) {
          await dispatch(
            select(global.localStorage.getItem('last-apd-id'), {
              global,
              pushRoute
            })
          );
        }
      })
      .catch(error => {
        const reason = error.response ? error.response.data : 'N/A';
        dispatch({ type: FETCH_ALL_APDS_FAILURE, error: reason });
      });
  };

export const deleteApd =
  (id, { fetch = fetchAllApds } = {}) =>
  dispatch => {
    dispatch({ type: DELETE_APD_REQUEST });

    return axios
      .delete(`/apds/${id}`)
      .then(() => {
        dispatch({ type: DELETE_APD_SUCCESS });
        dispatch(fetch());
      })
      .catch(() => {
        dispatch({ type: DELETE_APD_FAILURE });
      });
  };

// eslint-disable-next-line react/display-name
export const toggleAdminCheck = value => dispatch => {
  dispatch({ type: ADMIN_CHECK_TOGGLE, data: value });
  return null;
};
// eslint-disable-next-line react/display-name
export const toggleMiniCheck = value => dispatch => {
  dispatch({ type: ADMIN_CHECK_COLLAPSE_TOGGLE, data: value });
  return null;
};
// eslint-disable-next-line react/display-name
export const toggleAdminCheckComplete = value => dispatch => {
  dispatch({ type: ADMIN_CHECK_COMPLETE_TOGGLE, data: value });
  return null;
};
