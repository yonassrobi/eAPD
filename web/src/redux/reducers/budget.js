import { RESET } from '../actions/app';
import { LOAD_BUDGET } from '../actions/budget';
import { defaultBudgetObject } from '@cms-eapd/common';

// eslint-disable-next-line default-param-last
const reducer = (state = defaultBudgetObject([]), action) => {
  switch (action.type) {
    case RESET:
      return {};
    case LOAD_BUDGET:
      return action.budget;
    default:
      return state;
  }
};

export default reducer;
