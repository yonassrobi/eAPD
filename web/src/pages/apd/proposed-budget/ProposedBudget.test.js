import React from 'react';
import { renderWithConnection, screen, waitFor } from 'apd-testing-library';
import apd from '../../../fixtures/ak-apd.json';
import budget from '../../../fixtures/ak-budget.json';
import mmisApd from '../../../fixtures/ak-apd-mmis.json';
import mmisBudget from '../../../fixtures/ak-budget-mmis.json';

import ProposedBudget from './ProposedBudget';

const hitechApd = {
  initialState: {
    ...apd,
    ...budget
  }
};

const mmisApdProp = {
  initialState: {
    ...mmisApd,
    ...mmisBudget
  }
};

const setup = (props = {}, options = {}) => {
  return renderWithConnection(<ProposedBudget {...props} />, options);
};

describe('<ProposedBudget />', () => {
  it('render correctly hitech', async () => {
    setup({}, hitechApd);

    await waitFor(() => {
      expect(screen.getAllByRole('table')).toHaveLength(32);
    });

    expect(screen.getByRole('heading', { name: 'Proposed Budget' }));
    expect(screen.getByRole('heading', { name: 'Combined Activity Costs' }));
    expect(screen.getByRole('heading', { name: 'Summary Budget Table' }));
    expect(screen.getByRole('heading', { name: 'Quarterly Federal Share' }));
    expect(
      screen.getByRole('heading', {
        name: 'Estimated Quarterly Incentive Payments'
      })
    );
  });

  it.skip('render correctly mmis', () => {
    setup({}, mmisApdProp); // TODO: Skipping for now since Proposed Budget page still crashes
  });
});
