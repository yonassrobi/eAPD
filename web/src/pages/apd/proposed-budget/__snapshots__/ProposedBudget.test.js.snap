// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposedBudget component renders correctly 1`] = `
<Fragment>
  <Connect(ConnectedWaypoint) />
  <Connect(AlertMissingFFY) />
  <Section
    id={null}
    resource="proposedBudget"
  >
    <Connect(ConnectedWaypoint)
      id="combined-activity-costs-table"
    />
    <Subsection
      headerClassName={null}
      id="combined-activity-costs-table"
      nested={false}
      resource="proposedBudget.combinedActivityCosts"
    >
      <Connect(CombinedActivityCosts) />
    </Subsection>
    <Connect(ConnectedWaypoint)
      id="budget-summary-table"
    />
    <Subsection
      headerClassName={null}
      id="budget-summary-table"
      nested={false}
      resource="proposedBudget.summaryBudget"
    >
      <Connect(BudgetSummary) />
    </Subsection>
    <Connect(ConnectedWaypoint)
      id="budget-federal-by-quarter"
    />
    <Subsection
      headerClassName={null}
      id="budget-federal-by-quarter"
      nested={false}
      resource="proposedBudget.quarterlyBudget"
    >
      <Connect(QuarterlyBudgetSummary) />
    </Subsection>
    <Connect(ConnectedWaypoint)
      id="budget-incentive-by-quarter"
    />
    <Subsection
      headerClassName={null}
      id="budget-incentive-by-quarter"
      nested={false}
      resource="proposedBudget.paymentsByFFYQuarter"
    >
      <Connect(IncentivePayments) />
    </Subsection>
  </Section>
</Fragment>
`;
